### PR TITLE
[WASMFS] rename syscall

### PR DESCRIPTION
**Pull request by ethanalee**
Relevant Issue: #15041 
- Implement rename
- Add to existing `test_rename.c`
- Gave write permissions to root directory.
- Introduce `recursive_mutex` and try locking mechanism.
- TODO: Need to refactor path parsing, clean up comments.

### CODE REVIEW

File: system/lib/wasmfs/file.cpp [hunk 1]
```diff
@@ -83,6 +85,15 @@ std::shared_ptr<Directory> getDir(std::vector<std::string>::iterator begin,
...
     curr = directory->locked().getEntry(*it);
 
+    // For the rename syscall one must check if the source dir is an ancestor of
+    // the destination dir.
+    if (ancestor) {
+      if (curr == ancestor) {
+        err = -EINVAL;
+        return nullptr;
+      }
+    }
```
Comment (kripken) [inline system/lib/wasmfs/file.cpp:95]: It looks like "ancestor" is something we stop and error at. How about calling it `stopAt` or `errorOn`?

File: system/lib/wasmfs/file.h [hunk 2]
```diff
@@ -79,10 +82,19 @@ class File : public std::enable_shared_from_this<File> {
 
   Handle locked() { return Handle(shared_from_this()); }
 
+  std::optional<Handle> maybeLocked() {
+    auto handle = Handle(shared_from_this(), std::defer_lock);
+    if (handle.trylock()) {
+      return Handle(shared_from_this());
+    } else {
+      return std::nullopt;
```
Comment (kripken) [inline system/lib/wasmfs/file.h:90]: Can this be `return {}`?

File: system/lib/wasmfs/syscalls.cpp [hunk 3]
```diff
@@ -719,4 +719,142 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  // 4. b) If the new path child does not exist, add the old path child to the
+  // new path parent.
+  // 4. c) If the new path child does exist, unlink the new path child and add
+  // the old path child. 
+  // Renaming a directory:
+  // 5. a) If the old path child is a directory, ensure the new path child is
+  // not a file.
+  // 5. b) If the new path child does not exist, do the same as 4b.
+  // 5. c) If the new path child does exist, ensure the newpath child is empty
+  // and then do 4c.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:745]: This level of detail in the comment worries me. It's like writing out a spec and then the code implements that spec. But as the spec is separate from the code, it would be very easy for future bugfixes to fix the code but not remember to press pageup and find the spec that needs updating.

Instead, I think an overview can go here, but then the details can go alongside the  code below?

File: system/lib/wasmfs/syscalls.cpp [hunk 4]
```diff
@@ -719,4 +719,142 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  if (!oldParentDir) {
+    return err;
+  }
+
+  // For this operation to be atomic we must lock the source parent directory.
+  auto lockedOldParentDir = oldParentDir->locked();
+
+  auto oldPath = lockedOldParentDir.getEntry(oldBase);
+
+  // Check if the old_path exists.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:778]: This comment doesn't add anything on the code below it, which is readable enough I think.

File: system/lib/wasmfs/file.cpp (hunk 1 above)
Comment (tlively) [reply system/lib/wasmfs/file.cpp:95]: or `forbiddenAncestor`?

File: system/lib/wasmfs/syscalls.cpp (hunk 3 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:745]: +1, rather than describing the specific checks the code does, it would be more helpful here to explain what edge cases it needs to handle and potential problems it needs to avoid.

File: system/lib/wasmfs/file.cpp (hunk 1 above)
Comment (kripken) [reply system/lib/wasmfs/file.cpp:95]: Ah, nice, `forbiddenAncestor` is even clearer!

File: system/lib/wasmfs/file.cpp [hunk 5]
```diff
@@ -54,9 +54,11 @@ void MemoryFile::Handle::preloadFromJS(int index) {
 //
 // Path Parsing utilities
 //
+
 std::shared_ptr<Directory> getDir(std::vector<std::string>::iterator begin,
                                   std::vector<std::string>::iterator end,
-                                  long& err) {
+                                  long& err,
+                                  std::shared_ptr<File> ancestor) {
```
[the remark this answers is not in the document]
Comment (ethanalee) [reply system/lib/wasmfs/file.cpp:61]: Yep I clarified that it should stop parsing when this is detected in the path.

File: tests/stdio/test_rename.c [hunk 6]
```diff
@@ -124,6 +131,37 @@ void test() {
   err = rename("dir/subdir4/", "dir/subdir5/");
   assert(!err);
 
+  // Test renaming the same directory
+  // In Linux this does nothing, but in the JS file system it reports ENOENT.
```
Comment (tlively) [inline tests/stdio/test_rename.c]:
```suggestion
  // In Linux and WasmFS this does nothing, but in the JS file system it reports ENOENT.
```

**Review by tlively — COMMENTED** [review]
Looks pretty good overall!

For the `cleanup` in the test, it might be better to programmatically traverse and delete the file tree from the main entry point (`dir`) rather than trying to list every file. Assuming `readdir` and `unlink` work correctly, that will be more robust than the current strategy of trying to list everything.

It would also be good to test more of the edge cases we came up with during development. The one that comes to mind is the one where the parent of the source appears in the path of the destination

File: system/lib/wasmfs/syscalls.cpp [hunk 7]
```diff
@@ -719,4 +719,142 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  // 5. a) If the old path child is a directory, ensure the new path child is
+  // not a file.
+  // 5. b) If the new path child does not exist, do the same as 4b.
+  // 5. c) If the new path child does exist, ensure the newpath child is empty
+  // and then do 4c.
+
+  // If old_path and new_path are the same, do nothing.
+  if (strcmp((char*)old_path, (char*)new_path) == 0) {
+    return 0;
+  }
```
[the remark this answers is not in the document]
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp]: Oh I changed it before this comment!

File: tests/stdio/test_rename.c [hunk 8]
```diff
@@ -124,6 +133,50 @@ void test() {
...
+  assert(errno == EBUSY);
+#else
+  assert(errno == EINVAL);
+#endif
+
+  err = rename("dir/file2", "/");
+  assert(err == -1);
+  assert(errno == ENOTEMPTY);
+
+  // Test renaming the current working directory while still root.
```
Comment (tlively) [inline tests/stdio/test_rename.c:154]: It would be nice to have a separate test file just for testing interactions between `chdir`, `getcwd`, `rename`, and `unlink`. Maybe we can just leave a TODO for that in this PR?

File: system/lib/wasmfs/syscalls.cpp [hunk 9]
```diff
@@ -719,4 +719,142 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+      if (newPathDirectory->locked().getNumEntries() > 0) {
+        return -ENOTEMPTY;
+      }
+
+      lockedNewParentDir.unlinkEntry(newBase);
+    }
+
+    // Unlink the oldpath and add the oldpath to the new parent dir.
+    lockedOldParentDir.unlinkEntry(oldBase);
+    lockedNewParentDir.setEntry(newBase, oldDirectory);
```
[the remark this answers is not in the document]
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:849]: Yep, that sounds right.

File: tests/stdio/test_rename.c (hunk 8 above)
Comment (ethanalee) [reply tests/stdio/test_rename.c:154]: Sure, do you want me to remove `chdir` and `getcwd` then?

File: system/lib/wasmfs/syscalls.cpp [hunk 10]
```diff
@@ -719,4 +719,136 @@ long __syscall_getdents64(long fd, long dirp, long count) {
 
   return bytesRead;
 }
+
+long __syscall_rename(long old_path, long new_path) {
+  // The rename syscall must be atomic to prevent other file system operations
+  // from changing it. For example, two renames could be contending for the same
+  // source and destination file. Locking both the old_path parent and new_path
+  // parent are necessary. One can attempt to trylock on the new_path parent.
+  // This should return EBUSY if new_path parent is already locked.
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:728]: > The rename syscall must be atomic to prevent other file system operations from changing it.

It would be good to be more specific about what "it" is here. What exactly do we need to prevent other file system operations from changing?

> For example, two renames could be contending for the same source and destination file. Locking both the old_path parent and new_path parent are necessary.

It would be good to join these sentences with ", so".

> One can attempt to trylock on the new_path parent. This should return EBUSY if new_path parent is already locked.

It would be good to explain why we need to trylock in the first place. Specifically, it would be good to explain that two threads could be trying to do renames withe opposing source and destination directories, which causes them to lock the directories in the reverse order, causing deadlock unless we trylock here.

File: system/lib/wasmfs/syscalls.cpp [hunk 11]
```diff
@@ -719,4 +719,142 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+
+  if (oldPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming the root directory returns EBUSY.
+  // TODO: Fix this when path parsing is refactored.
+  if (oldPathParts.size() == 1 && oldPathParts[0] == "/") {
+    return -EBUSY;
+  }
```
[the remark this answers is not in the document]
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:746]: This needs to wait for the path parsing refactoring. If we changed it now it would conflict with the case where `parent` == `root` and `child` == `null`.

File: system/lib/wasmfs/syscalls.cpp [hunk 12]
```diff
@@ -722,10 +722,14 @@ long __syscall_getdents64(long fd, long dirp, long count) {
 
 long __syscall_rename(long old_path, long new_path) {
   // The rename syscall must be atomic to prevent other file system operations
-  // from changing it. For example, two renames could be contending for the same
-  // source and destination file. Locking both the old_path parent and new_path
-  // parent are necessary. One can attempt to trylock on the new_path parent.
-  // This should return EBUSY if new_path parent is already locked.
+  // from changing it. If it were not atomic, then another thread could unlink
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp]:
```suggestion
  // from concurrently changing the directories. If it were not atomic, then another thread could unlink
```

File: system/lib/wasmfs/syscalls.cpp [hunk 13]
```diff
@@ -722,10 +722,14 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  // the destination directory after we have moved our source child. This would
+  // leave the file system in an undefined state as the source child is now
+  // orphaned. Protecting the destination and source parent directories ensures
+  // that no resources that we depend on are altered. Trylocking on the new_path
+  // parent is needed in the case where two renames are operating on opposing
+  // sources and destinations. This could cause them to lock the directories in
+  // reverse order and cause deadlock.
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp]:
```suggestion
  // sources and destinations. This would cause them to lock the directories in
  // reverse order and could cause deadlock.
```

File: system/lib/wasmfs/syscalls.cpp [hunk 14]
```diff
@@ -719,4 +719,133 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+
+    // This should also cover the case in where
+    // the destination is an ancestor of the source:
+    // rename("dir/subdir", "dir");
+    if (newPathDirectory->locked().getNumEntries() > 0) {
+      return -ENOTEMPTY;
+    }
+
+    lockedNewParentDir.unlinkEntry(newBase);
+  }
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:843]: There is still some unnecessary duplication here: `newPath` is checked twice and unlinked twice. This structure could fix it:

```
if (newPath) {
  if (auto oldFile = oldPath->dynCast<DataFile>()) {
    // checks...
  } else if (auto oldDir = oldPath->dynCast<Directory>()) {
    // checks...
  } else {
    assert(false && "unhandled file kind");
  }
  lockedNewParentDir.unlinkEntry(newBase);
}
```

**Review by tlively — APPROVED** [review]
LGTM with that last comment :)

File: system/lib/wasmfs/syscalls.cpp (hunk 14 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:843]: Hmm I realize that at the end we don't even really need to casted file objects - `lockedNewParentDir.setEntry(newBase, oldPath);`

We can probably just drop this `auto oldDir = oldPath->dynCast<Directory>()`, etc.

File: system/lib/wasmfs/file.h [hunk 15]
```diff
@@ -241,9 +264,12 @@ class MemoryFile : public DataFile {
 
 // Obtains parent directory of a given pathname.
 // Will return a nullptr if the parent is not a directory.
-std::shared_ptr<Directory> getDir(std::vector<std::string>::iterator begin,
-                                  std::vector<std::string>::iterator end,
-                                  long& err);
+// Will exit if the forbiddenAncestor shared_ptr is encountered while parsing.
```
[the remark this answers is not in the document]
Comment (kripken) [reply system/lib/wasmfs/file.h:267]: Also please mention that it not just exits but also errors, and how it errors.

File: system/lib/wasmfs/file.cpp [hunk 16]
```diff
@@ -83,6 +85,17 @@ std::shared_ptr<Directory> getDir(std::vector<std::string>::iterator begin,
 #endif
     curr = directory->locked().getEntry(*it);
 
+    // Special case for the rename syscall.
+    // If one detects that old_path is a forbidden ancestor in the path of
+    // new_path, the rename operation is invalid. Ex.
+    // rename("dir", "dir/somename").
```
Comment (kripken) [inline system/lib/wasmfs/file.cpp:91]: This comment is in high danger of being stale and misleading. Commenting on the internal details of distant code is generally not a good idea. It does increase readability of this PR's diff, but the more important thing is all the reading of the code that will happen in the future. Those people will read this file by itself.

I don't think a comment is needed here: forbiddenAncestor is explained in the header, and this just does what that says.

File: system/lib/wasmfs/file.h [hunk 17]
```diff
@@ -79,10 +82,19 @@ class File : public std::enable_shared_from_this<File> {
...
+    } else {
+      return {};
+    }
+  }
+
 protected:
   File(FileKind kind, mode_t mode) : kind(kind), mode(mode) {}
   // A mutex is needed for multiple accesses to the same file.
-  std::mutex mutex;
+  std::recursive_mutex mutex;
```
Comment (kripken) [inline system/lib/wasmfs/file.h:97]: When we have benchmarks it would be good to see what effect this has (we will be able to flip this to a normal mutex so long as the benchmark doesn't rename).

**Review by tlively — APPROVED** [review]
Great!

File: system/lib/wasmfs/syscalls.cpp [hunk 18]
```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
 
   return bytesRead;
 }
+
+long __syscall_rename(long old_path, long new_path) {
+  // The rename syscall must be atomic to prevent other file system operations
+  // from concurrently changing the directories. If it were not atomic, then
+  // other threads could unlink the destination directory after we have moved
+  // our source child. This would leave the file system in an undefined state as
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:727]: I'm not sure this is the best example. Unlinking the new path, and orphaning the file, is not that bad: imagine that the unlink happened right after the rename, then the result would be the same. That the file is moved to an orphaned directory is still a consistent state, I think.

An example of actually inconsistent state would be if the file ends up in two places, or has two parents or zero parents. That is, inconsistent state is something that a "filesystem validator" could flag as being obviously wrong. But a file that is unlinked is valid.

File: system/lib/wasmfs/syscalls.cpp [hunk 19]
```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+
+  auto oldPathParts = splitPath((char*)old_path);
+
+  if (oldPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming the root directory returns EBUSY.
+  // TODO: Fix this when path parsing is refactored.
+  if (oldPathParts.size() == 1 && oldPathParts[0] == "/") {
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:748]: Can this be written `oldPathParts == { "/" }`? (I have no idea, but maybe... :smile: )

Another option might be a helper `isRootDir` perhaps.

File: system/lib/wasmfs/syscalls.cpp [hunk 20]
```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  }
+
+  // Obtain the destination parent directory to see if it exists.
+  auto newPathParts = splitPath((char*)new_path);
+
+  if (newPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming a directory with the root directory returns ENOTEMPTY.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:777]:
```suggestion
  // In Linux, renaming a directory to the root directory returns ENOTEMPTY.
```

File: system/lib/wasmfs/syscalls.cpp [hunk 21]
```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  // TODO: Fix this when path parsing is refactored.
+  if (newPathParts.size() == 1 && newPathParts[0] == "/") {
+    return -ENOTEMPTY;
+  }
+
+  auto newBase = newPathParts.back();
+
+  auto newParentDir =
+    getDir(newPathParts.begin(), newPathParts.end() - 1, err, oldPath);
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:786]: A comment here about the use of oldPath would be useful, as it is not actually clear while reading this that oldPath is assigned to forbiddenAncestor (that name would have clarified things).

File: system/lib/wasmfs/file.h (hunk 17 above)
Comment (kripken) [reply system/lib/wasmfs/file.h:97]: A non-trivial comment explaining why this mutex is recursive is necessary here, I think.

File: system/lib/wasmfs/syscalls.cpp [hunk 22]
```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+
+  auto lockedNewParentDir = std::move(*maybeLockedNewParentDir);
+  auto newPath = lockedNewParentDir.getEntry(newBase);
+
+  // If old_path and new_path are the same, do nothing.
+  if (newPath == oldPath) {
+    return 0;
+  }
+
+  // Cannot move to a destination parent directory wihout write permissions.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:810]:
```suggestion
  // Cannot move to a destination parent directory without write permissions.
```

File: system/lib/wasmfs/syscalls.cpp [hunk 23]
```diff
@@ -719,4 +719,130 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+  if (!(lockedNewParentDir.mode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:813]: Don't we need the same permission on the old path? We are modifying that directory as well.

File: system/lib/wasmfs/file.cpp (hunk 16 above)
Comment (ethanalee) [reply system/lib/wasmfs/file.cpp:91]: That sounds good. I was originally trying to enhance readability and also thought that since `getDir` is an internal function, any re-factoring would probably change this area of the code anyways.

File: system/lib/wasmfs/syscalls.cpp (hunk 19 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:748]: I can do:
```
std::vector<std::string> root = {"/"};
  if (oldPathParts == root) {
    return -EBUSY;
  }
```

File: system/lib/wasmfs/syscalls.cpp (hunk 23 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:813]: Hmm good point, this is a good case to add to the tests as well.

File: system/lib/wasmfs/syscalls.cpp (hunk 18 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp]: I remember you provided an example earlier. I included here. Wondering if it works here @kripken ?
